File: templater/flags.py
import argparse
import logging as log
import os
import sys

import templater

parser = None
args = None


def parse(argv=None):
    global parser
    global args

    parser = argparse.ArgumentParser(
        prog="template-dockerfiles",
        description="A Docker image builder that uses Jinja2 templates to dynamically generate Dockerfiles.",
        epilog="When 'docker build' is just not enough :-)",
    )
    parser.add_argument(
        "-c",
        "--config",
        action="store",
        dest="CONFIG_FILE",
        help="path to the configuration file",
        type=is_valid_file,
        required=True,
    )

    parser.add_argument(
        "--dry-run",
        help="print what would be done, but don't do anything",
        action="store_const",
        dest="DRY_RUN",
        const=True,
        default=False,
    )

    parser.add_argument(
        "--push",
        help="push Docker images when successfully build",
        action="store_const",
        dest="PUSH",
        const=True,
        default=False,
    )

    parser.add_argument(
        "--parallel",
        help="specify the number of threads to use (default: number of CPUs).",
        type=validate_threads,
        dest="THREADS",
        default=1,
    )

    parser.add_argument(
        "-v",
        "--verbose",
        help="be verbose",
        action="store_const",
        dest="LOG_LEVEL",
        const=log.DEBUG,
        default=log.INFO,
    )

    parser.add_argument(
        "--version",
        help="show the version of the application and exit",
        action="version",
        version=f"%(prog)s {templater.__version__}",
    )

    # parser.add_argument('--debug',
    #                     help="Print lots of debugging statements",
    #                     action="store_const",
    #                     dest="LOG_LEVEL",
    #                     const=log.DEBUG)

    parser.add_argument(
        "-t",
        "--tag",
        action="store",
        dest="TAG",
        help="tag that could be used as an image version",
        required=True,
        default=None,
    )

    # parse only once
    if args is None:
        args = vars(parser.parse_args(argv or sys.argv[1:]))

    return args


def validate_threads(value):
    """Validate that the provided thread count is a positive integer."""
    if value.lower() == "max":
        return "max"
    try:
        threads = int(value)
        if threads < 1:
            raise ValueError
        return threads
    except ValueError:
        raise argparse.ArgumentTypeError(
            f"Invalid value for threads: {value}. Must be 'max' or a positive integer."
        )


def is_valid_file(file_path):
    if not os.path.isfile(file_path):
        raise argparse.ArgumentTypeError(
            f"The file '{file_path}' does not exist or is not a valid file."
        )
    return file_path


def __getattr__(name):
    global args
    if args is None:
        raise RuntimeError("Arguments have not been parsed yet.")
    if name in args.keys():
        return args[name]
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")


def all():
    global args
    if args is None:
        raise RuntimeError("Arguments have not been parsed yet.")
    return args


def add(name, value):
    global args
    args.update({name: value})

File: pkg/flags/flags.go
package flags

import "flag"
import (
	"fmt"
	"os"
	"runtime"

	"github.com/spf13/cobra"
)

type Args struct {
    ConfigFile 	string
    Verbose    	bool
	DryRun		bool
	Push		bool
	Threads		int
	//LogLevel	string
	Version		bool
	Tag			string

	// ConfigFile string `short:"c" long:"config" description:"Path to the configuration file" required:"true"`
	// DryRun     bool   `long:"dry-run" description:"Print what would be done, but don't do anything"`
	// Push       bool   `long:"push" description:"Push Docker images when successfully built"`
	// Threads    int    `long:"parallel" description:"Specify the number of threads to use (default: number of CPUs)" default:"1"`
	// LogLevel   string `short:"v" long:"verbose" description:"Set log level" choice:"DEBUG" choice:"INFO" default:"INFO"`
	// Version    bool   `long:"version" description:"Show the version of the application and exit"`
	// Tag        string `short:"t" long:"tag" description:"Tag that could be used as an image version" required:"true"`
}

var rootCmd = &cobra.Command{
	Use:   "template-dockerfiles",
	Short: "A Docker image builder that uses Go templates to dynamically generate Dockerfiles.",
	Long:  `A CLI tool for building Docker images with configurable Dockerfile templates and multi-threaded execution.`,
	Run: func(cmd *cobra.Command, args []string) {
		// Main logic goes here
		if verbose {
			fmt.Println("Verbose mode enabled")
		}
		fmt.Printf("Using config file: %s\n", configFile)
	},
}


func Parse() *Args {
    var args Args
    // flag.StringVar(&args.ConfigFile, "config", "build.yaml", "Path to configuration file")
    // flag.BoolVar(&args.Verbose, "verbose", false, "Enable verbose logging")
    // flag.Parse()

	rootCmd.PersistentFlags().StringVarP(&args.ConfigFile, "config", "c", "", "Path to the configuration file (required)")
	rootCmd.MarkPersistentFlagRequired("config")

	rootCmd.Flags().BoolVar(&args.DryRun, "dry-run", false, "Print actions but don't execute them")
	rootCmd.Flags().BoolVar(&args.Push, "push", false, "Push Docker images after building")
	rootCmd.Flags().IntVarP(&args.Threads, "parallel", "p", runtime.NumCPU(), "Specify the number of threads to use (default: number of CPUs)")
	rootCmd.Flags().StringVarP(&args.Tag, "tag", "t", "", "Tag to use as the image version")
	rootCmd.Flags().BoolVarP(&args.Verbose, "verbose", "v", false, "Increase verbosity of output")
	rootCmd.Flags().BoolP("version", "V", false, "Display the application version and exit")

    return &args
}






func main() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
